janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/ArrayUtil.java

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.util;
+
+public class ArrayUtil {
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    /**
+     * Given an old capacity of an array, generate the new capacity that is larger than or equal to minimum required capacity.
+     * Under most circumstances, it doubles the old capacity. If the old capacity is already larger than half of @{link #MAX_ARRAY_SIZE},
+     * then new capacity will be {@link #MAX_ARRAY_SIZE}, which is slightly smaller than Integer.MAX_VALUE to avoid
+     * Out of Memory error. If the min capacity is smaller than old capacity, then old capacity will be simply returned.
+     * @param oldCapacity old capacity
+     * @param minCapacity minimum desired capacity
+     * @return a new capacity that is larger than or equal to minCapacity
+     * @throws IllegalArgumentException if minimum required capacity is larger than {@link #MAX_ARRAY_SIZE}
+     */
+    public static int growSpace(int oldCapacity, int minCapacity) {
+        if (minCapacity < oldCapacity) {
+            return oldCapacity;
+        }
+        if (minCapacity > MAX_ARRAY_SIZE) {
+            throw new IllegalArgumentException(String.format("required array capacity %d is larger than MAX_ARRAY_SIZE [%d]",
+                minCapacity, MAX_ARRAY_SIZE));
+        }
+        int newCapacity = oldCapacity << 1;
+        if (newCapacity - minCapacity < 0) {
+            newCapacity = minCapacity;
+        }
+        if (newCapacity - MAX_ARRAY_SIZE > 0) {
+            newCapacity = MAX_ARRAY_SIZE;
+        }
```

Comment:
As far as I understand, these two if statements can be chained together via `else` because `minCapacity` here cannot be bigger than `MAX_ARRAY_SIZE`. I.e. 
```
if (newCapacity - minCapacity < 0) {
    newCapacity = minCapacity;
} else if (newCapacity - MAX_ARRAY_SIZE > 0) {
    newCapacity = MAX_ARRAY_SIZE;
}
```

Also, I'm really surprised to see such checks implementations (i.e. `newCapacity - minCapacity < 0` instead of `newCapacity < minCapacity`). I see that this implementation is used in OpenJDK 7 and 8. Probably it's just some optimizations which I'm not aware of.

Reply:
I wasn't aware of this optimization either. But when I think about it, it makes sense to me.
`a - b < 0 == a < b` only holds for positive integers (counter example: `a = Integer.MAX_VALUE` and `b = -1`) so when Java evaluates `a < b` internally using subtraction, it also needs to check whether `a` and `b` are positive or negative in order to interpret the result.
As we know in our case that we can only have positive integers (do we actually know?), this check can be simplified. Disclaimer: This is just what came to my mind and I didn't find any source that proves or disproves it.

Reply:
You could check out my branch, play around with it locally and run the test :P  There are some tricks here. At the first glance, some comparisons here seem weird or even redundant, but they are actually not.

Seems the if statements can be chained together - I will double confirm and update the PR. Thanks for noticing that!

The `newCapacity - minCapacity < 0` is necessary. `newCapacity < minCapacity` does not work properly because `newCapacity` might have already overflowed and is interpreted as a negative number. In this case, `newCapacity - minCapacity` will cause integer underflow and thus `newCapacity - minCapacity < 0` evaluates to false, while `newCapacity < minCapacity` evaluates to true.

Reply:
When `oldCapacity` is equals to `MAX_ARRAY_SIZE*0.7` and `minCapacity` is equals to `MAX_ARRAY_SIZE+1` what should this method return?
As I'm reading the code (didn't check it yet) it looks like this method will return `MAX_ARRAY_SIZE+1` but from my understanding this method should return `MAX_ARRAY_SIZE` because we doubling the size of `oldCapacity` and when it overflows we just return the maximum possible array. What do you think? Or do you think it's better to increment the array size slowly after it reaches the size of `(MAX_ARRAY_SIZE+9) / 2`?

Reply:
> When oldCapacity is equals to MAX_ARRAY_SIZE*0.7 and minCapacity is equals to MAX_ARRAY_SIZE*0.7+1 what should this method return

It returns MAX_ARRAY_SIZE. What happens is, `int newCapacity = oldCapacity << 1` will overflow, and then `newCapacity - minCapacity` will underflow thus `if (newCapacity - minCapacity < 0)` evaluates to false, while `newCapacity - MAX_ARRAY_SIZE > 0` evaluates to true and then we have `newCapacity = MAX_ARRAY_SIZE`.

Reply:
Now I see it. Thank you for the explanation!